covid19/spreadsheet_validator.py

```diff
@@ -0,0 +1,100 @@
+INVALID_CITY_CODE = -1
+
+
+def format_spreadsheet_rows_as_dict(rows_table, date, state):
+    """
+    Receives rows.Table object, a date and a brazilan UF, validates the data
+    and returns a dict in the formatted data:
+
+    This is an auxiliary method used by covid19.forms.StateSpreadsheetForm with the uploaded file
+    """
+    # TODO: https://github.com/turicas/brasil.io/issues/210
+    field_names = rows_table.field_names
+
+    confirmed_attr = _get_column_name(
+        field_names, ['confirmados', 'confirmado', 'casos_confirmados']
+    )
+    deaths_attr = _get_column_name(field_names, ['obitos', 'obito', 'morte', 'mortes'])
+    city_attr = _get_column_name(field_names, ['municipio', 'cidade'])
+
+    if not rows_table.fields[confirmed_attr] == IntegerField:
+        raise ValidationError('A coluna "Confirmados" precisa ter somente números inteiros"')
+    if not rows_table.fields[deaths_attr] == IntegerField:
+        raise ValidationError('A coluna "Mortes" precisa ter somente números inteiros"')
+
+    results = []
+    bad_cities = []
+    has_total, has_undefined = False, False
+    for i, entry in enumerate(rows_table):
+        city = getattr(entry, city_attr, None)
+        confirmed = getattr(entry, confirmed_attr, None)
+        deaths = getattr(entry, deaths_attr, None)
+
+        if confirmed is None:
+            raise ValidationError(f'Valor nulo para casos confirmados na linha {i + 1} da planilha')
+        if deaths is None:
+            raise ValidationError(f'Valor nulo para óbitos na linha {i + 1} da planilha')
+        if deaths > confirmed:
+            raise ValidationError(
+                f'Valor de óbitos maior que casos confirmados na linha {i + 1} da planilha'
+            )
+
+        result = _parse_city_data(city, confirmed, deaths, date, state)
+        if result['city_ibge_code'] == INVALID_CITY_CODE:
+            bad_cities.append(city)
+        elif not has_total and result['city'] == TOTAL_LINE_DISPLAY:
+            has_total = True
+        elif not has_undefined and result['city'] == UNDEFINED_DISPLAY:
+            has_undefined = True
+        results.append(result)
+
+    if not has_total:
+        raise ValidationError(f'A linha "{TOTAL_LINE_DISPLAY}" está faltando na planilha')
+    if not has_undefined:
+        raise ValidationError(f'A linha "{UNDEFINED_DISPLAY}" está faltando na planilha')
+    if bad_cities:
+        bad_cities = ', '.join(bad_cities)
+        raise ValidationError(f'As seguintes cidades não pertencem a UF {state}: {bad_cities}')
+
+    return results
+
+
+def _parse_city_data(city, confirmed, deaths, date, state):
+    data = {
+        "city": city,
+        "confirmed": confirmed,
+        "date": date.isoformat(),
+        "deaths": deaths,
+        "place_type": "city",
+        "state": state,
+    }
+
+    if city == TOTAL_LINE_DISPLAY:
+        data['city_ibge_code'] = get_state_info(state).state_ibge_code
+        data['place_type'] = 'state'
```

Comment:
Nesse caso, forçar `data["city"] = None`.

Reply:
@turicas eu segui a referência do script [`corona_sp_spider.py`](https://github.com/turicas/covid19-br/blob/master/corona_sp_spider.py#L80). Só queria checar explicitamente se o script está desatualizado e é de fato para ficar `None` no caso do total.

Reply:
Opa, o script está errado. TOTAL NO ESTADO só deve aparecer na planilha (seja na que o voluntário enviará por upload ou na do Google Spreadsheets que usamos hoje). Nessa versão já convertida é `None` mesmo. Vou consertar. Obrigado!

Reply:
Resolvido em 7343d8c